File: binding-native/src/c/sk_runtimeeffect.cpp
/*
 * Copyright 2020 Microsoft Corporation. All rights reserved.
 *
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

#include "include/core/SkTypes.h"
#include "include/core/SkColorFilter.h"
#include "include/core/SkShader.h"
#include "include/effects/SkRuntimeEffect.h"

#include "include/c/sk_types.h"
#include "include/c/sk_runtimeeffect.h"

#include "src/c/sk_types_priv.h"

// sk_runtimeeffect_t

sk_runtimeeffect_t* sk_runtimeeffect_make_for_shader(sk_string_t* sksl, sk_string_t* error) {
    auto [effect, errorMessage] = SkRuntimeEffect::MakeForShader(AsString(*sksl));
    if (error && errorMessage.size() > 0)
        AsString(error)->swap(errorMessage);
    return ToRuntimeEffect(effect.release());
}

sk_runtimeeffect_t* sk_runtimeeffect_make_for_color_filter(sk_string_t* sksl, sk_string_t* error) {
    auto [effect, errorMessage] = SkRuntimeEffect::MakeForColorFilter(AsString(*sksl));
    if (error && errorMessage.size() > 0)
        AsString(error)->swap(errorMessage);
    return ToRuntimeEffect(effect.release());
}

sk_runtimeeffect_t* sk_runtimeeffect_make_for_blender(sk_string_t* sksl, sk_string_t* error) {
    auto [effect, errorMessage] = SkRuntimeEffect::MakeForBlender(AsString(*sksl));
    if (error && errorMessage.size() > 0)
        AsString(error)->swap(errorMessage);
    return ToRuntimeEffect(effect.release());
}

sk_runtimeeffect_child_type_t sk_runtimeeffect_get_type(const sk_runtimeeffect_t* effect) {
    auto effect_ = AsRuntimeEffect(effect);
    return effect_->allowShader() ? SHADER_CHILD_TYPE : effect_->allowColorFilter() ? COLORFILTER_CHILD_TYPE : BLENDER_CHILD_TYPE;
}

void sk_runtimeeffect_unref(sk_runtimeeffect_t* effect) {
    SkSafeUnref(AsRuntimeEffect(effect));
}

sk_shader_t* sk_runtimeeffect_make_shader(sk_runtimeeffect_t* effect, sk_data_t* uniforms, sk_shader_t** children, size_t childCount, const sk_matrix_t* localMatrix) {
    sk_sp<SkShader>* skChildren = new sk_sp<SkShader>[childCount];
    for (int i = 0; i < childCount; i++) {
        skChildren[i] = sk_ref_sp(AsShader(children[i]));
    }

    SkMatrix m;
    if (localMatrix)
        m = AsMatrix(localMatrix);
    
    sk_sp<SkShader> shader = AsRuntimeEffect(effect)->makeShader(
        sk_ref_sp(AsData(uniforms)),
        skChildren,
        childCount,
        localMatrix ? &m : nullptr);

    delete[] skChildren;

    return ToShader(shader.release());
}

sk_colorfilter_t* sk_runtimeeffect_make_color_filter(sk_runtimeeffect_t* effect, sk_data_t* uniforms, sk_colorfilter_t** children, size_t childCount) {
    sk_sp<SkColorFilter>* skChildren = new sk_sp<SkColorFilter>[childCount];
    for (int i = 0; i < childCount; i++) {
        skChildren[i] = sk_ref_sp(AsColorFilter(children[i]));
    }

    sk_sp<SkColorFilter> shader = AsRuntimeEffect(effect)->makeColorFilter(
        sk_ref_sp(AsData(uniforms)),
        skChildren,
        childCount);

    delete[] skChildren;

    return ToColorFilter(shader.release());
}

sk_blender_t* sk_runtimeeffect_make_blender(sk_runtimeeffect_t* effect, sk_data_t* uniforms, sk_blender_t** children, size_t childCount) {
    SkRuntimeEffect::ChildPtr* skChildren = new SkRuntimeEffect::ChildPtr[childCount];
    for (int i = 0; i < childCount; i++) {
        skChildren[i] = SkRuntimeEffect::ChildPtr(sk_ref_sp(AsBlender(children[i])));
    }

    sk_sp<SkBlender> shader = AsRuntimeEffect(effect)->makeBlender(
        sk_ref_sp(AsData(uniforms)),
        SkSpan<SkRuntimeEffect::ChildPtr>(skChildren, childCount)
    );

    delete[] skChildren;

    return ToBlender(shader.release());
}

size_t sk_runtimeeffect_get_uniform_size(const sk_runtimeeffect_t* effect) {
    return AsRuntimeEffect(effect)->uniformSize();
}

size_t sk_runtimeeffect_get_uniform_count(const sk_runtimeeffect_t* effect) {
    return AsRuntimeEffect(effect)->uniforms().size();
}

const sk_runtimeeffect_uniform_t* sk_runtimeeffect_get_uniform_from_name(const sk_runtimeeffect_t* effect, const char* name) {
    return ToRuntimeEffectUniform(AsRuntimeEffect(effect)->findUniform(name));
}

const sk_runtimeeffect_uniform_t* sk_runtimeeffect_get_uniform_from_index(const sk_runtimeeffect_t* effect, int index) {
    SkSpan<const SkRuntimeEffect::Uniform> vector = AsRuntimeEffect(effect)->uniforms();
    auto item = vector.begin() + index;
    return ToRuntimeEffectUniform(&(*item));
}

const sk_runtimeeffect_child_t* sk_runtimeeffect_get_child_from_name(const sk_runtimeeffect_t* effect, const char* name) {
    return ToRuntimeEffectChild(AsRuntimeEffect(effect)->findChild(name));
}

const sk_runtimeeffect_child_t* sk_runtimeeffect_get_child_from_index(const sk_runtimeeffect_t* effect, int index) {
    SkSpan<const SkRuntimeEffect::Child> vector = AsRuntimeEffect(effect)->children();
    auto item = vector.begin() + index;
    return ToRuntimeEffectChild(&(*item));
}

size_t sk_runtimeeffect_get_children_count(const sk_runtimeeffect_t* effect) {
    return AsRuntimeEffect(effect)->children().size();
}

// sk_runtimeeffect_uniform_t

void sk_runtimeeffect_uniform_get_name(const sk_runtimeeffect_uniform_t* child, sk_string_t* name) {
    auto item = AsRuntimeEffectUniform(child);
    AsString(name)->set(item->name.data(), item->name.size());
}

int sk_runtimeeffect_uniform_get_offset(const sk_runtimeeffect_uniform_t* child) {
    return AsRuntimeEffectUniform(child)->offset;
}

sk_runtimeeffect_uniform_type_t sk_runtimeeffect_uniform_get_type(const sk_runtimeeffect_uniform_t* child) {
    return (sk_runtimeeffect_uniform_type_t)AsRuntimeEffectUniform(child)->type;
}

size_t sk_runtimeeffect_uniform_get_count(const sk_runtimeeffect_uniform_t* child) {
    return AsRuntimeEffectUniform(child)->count;
}

size_t sk_runtimeeffect_uniform_get_size_in_bytes(const sk_runtimeeffect_uniform_t* child) {
    return AsRuntimeEffectUniform(child)->sizeInBytes();
}

// sk_runtimeeffect_child_t

void sk_runtimeeffect_child_get_name(const sk_runtimeeffect_child_t* child, sk_string_t* name) {
    auto item = AsRuntimeEffectChild(child);
    AsString(name)->set(item->name.data(), item->name.size());
}

sk_runtimeeffect_child_type_t sk_runtimeeffect_child_get_type(const sk_runtimeeffect_child_t* child) {
    return (sk_runtimeeffect_child_type_t)AsRuntimeEffectChild(child)->type;
}

int sk_runtimeeffect_child_get_index(const sk_runtimeeffect_child_t* child) {
    return AsRuntimeEffectChild(child)->index;
}

File: binding/Binding/SKRuntimeEffect.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace SkiaSharp
{
	public unsafe class SKRuntimeEffect : SKObject, ISKReferenceCounted
	{
		public unsafe class Uniform : SKObject
		{
			internal Uniform (IntPtr handle, bool owns)
				: base (handle, owns)
			{
			}

			internal static Uniform GetObject (IntPtr handle) =>
				GetOrAddObject (handle, false, (h, o) => new Uniform (h, o));

			// Properties

			public string Name {
				get {
					using var str = new SKString ();
					SkiaApi.sk_runtimeeffect_uniform_get_name (Handle, str.Handle);
					return str.ToString ();
				}
			}

			public int Offset =>
				(int)SkiaApi.sk_runtimeeffect_uniform_get_offset (Handle);

			public SKRuntimeEffectUniformType Type =>
				SkiaApi.sk_runtimeeffect_uniform_get_type (Handle);

			public int Count =>
				(int)SkiaApi.sk_runtimeeffect_uniform_get_count (Handle);

			public int SizeInBytes =>
				(int)SkiaApi.sk_runtimeeffect_uniform_get_count (Handle);
		}

		public unsafe class Child : SKObject
		{
			internal Child (IntPtr handle, bool owns)
				: base (handle, owns)
			{
			}

			internal static Child GetObject (IntPtr handle) =>
				GetOrAddObject (handle, false, (h, o) => new Child (h, o));

			// Properties

			public string Name {
				get {
					using var str = new SKString ();
					SkiaApi.sk_runtimeeffect_child_get_name (Handle, str.Handle);
					return str.ToString ();
				}
			}

			public SKRuntimeEffectChildType Type =>
				SkiaApi.sk_runtimeeffect_child_get_type (Handle);

			public int Index =>
				(int)SkiaApi.sk_runtimeeffect_child_get_index (Handle);
		}

		internal SKRuntimeEffect (IntPtr handle, bool owns)
			: base (handle, owns)
		{
		}

		internal static SKRuntimeEffect GetObject (IntPtr handle) =>
			GetOrAddObject (handle, (h, o) => new SKRuntimeEffect (h, o));

		internal static SKRuntimeEffect GetObject (IntPtr handle, bool owns) =>
			GetOrAddObject (handle, owns, (h, o) => new SKRuntimeEffect (h, o));

		// Create

		public static SKRuntimeEffect CreateForShader (string sksl, out string errors)
		{
			using var s = new SKString (sksl);
			using var errorString = new SKString ();
			var effect = GetObject (SkiaApi.sk_runtimeeffect_make_for_shader (s.Handle, errorString.Handle));
			errors = errorString?.ToString ();
			if (errors?.Length == 0)
				errors = null;
			return effect;
		}

		public static SKRuntimeEffect CreateForColorFilter (string sksl, out string errors)
		{
			using var s = new SKString (sksl);
			using var errorString = new SKString ();
			var effect = GetObject (SkiaApi.sk_runtimeeffect_make_for_color_filter (s.Handle, errorString.Handle));
			errors = errorString?.ToString ();
			if (errors?.Length == 0)
				errors = null;
			return effect;
		}

		public static SKRuntimeEffect CreateForBlender (string sksl, out string errors)
		{
			using var s = new SKString (sksl);
			using var errorString = new SKString ();
			var effect = GetObject (SkiaApi.sk_runtimeeffect_make_for_blender (s.Handle, errorString.Handle));
			errors = errorString?.ToString ();
			if (errors?.Length == 0)
				errors = null;
			return effect;
		}

		// Properties

		public SKRuntimeEffectChildType EffectType =>
			SkiaApi.sk_runtimeeffect_get_type (Handle);

		public int UniformSize =>
			(int)SkiaApi.sk_runtimeeffect_get_uniform_size (Handle);

		public int ChildrenCount =>
			(int)SkiaApi.sk_runtimeeffect_get_children_count (Handle);

		public int UniformCount =>
			(int)SkiaApi.sk_runtimeeffect_get_uniform_count (Handle);

		public Uniform GetUniform (int i)
		{
			if (i < 0 || i >= ChildrenCount)
				return null;

			return Uniform.GetObject(SkiaApi.sk_runtimeeffect_get_uniform_from_index (Handle, i));
		}

		public Uniform GetUniform (string name)
		{
			return Uniform.GetObject (SkiaApi.sk_runtimeeffect_get_uniform_from_name (Handle, name));
		}

		public Child GetChild (int i)
		{
			if (i < 0 || i >= ChildrenCount)
				return null;

			return Child.GetObject (SkiaApi.sk_runtimeeffect_get_child_from_index (Handle, i));
		}

		public Child GetChild (string name)
		{
			return Child.GetObject (SkiaApi.sk_runtimeeffect_get_child_from_name (Handle, name));
		}
	}
}
